File: aws-ts-ec2-provisioners/provisioners.ts
// Copyright 2016-2019, Pulumi Corporation.  All rights reserved.

import * as pulumi from "@pulumi/pulumi";
import { Provisioner } from "./provisioner";

// ConnectionArgs tells a provisioner how to access a remote resource. For example, it may need to use
// SSH or WinRM to connect to the resource in order to run a command or copy a file.
export interface ConnectionArgs {
    type?: ConnectionType;
    host: pulumi.Input<string>;
    port?: pulumi.Input<number>;
    username?: pulumi.Input<string>;
    password?: pulumi.Input<string>;
    privateKey?: pulumi.Input<string>;
    privateKeyPassphrase?: pulumi.Input<string>;
}

// ConnectionType is the set of legal connection mechanisms to use. Default is SSH.
export type ConnectionType = "ssh" | "winrm";

function connTypeOrDefault(conn: pulumi.Unwrap<ConnectionArgs>): ConnectionType {
    return conn.type || "ssh";
}

function connPortOrDefault(conn: pulumi.Unwrap<ConnectionArgs>): number {
    if (conn.port !== undefined) {
        return conn.port;
    }

    const connType = connTypeOrDefault(conn);
    switch (connType) {
        case "ssh":
            return 22;
        case "winrm":
            return 5985;
        default:
            throw new Error(`unrecognized connectiont ype ${connType}`);
    }
}

function connUsernameOrDefault(conn: pulumi.Unwrap<ConnectionArgs>): string {
    if (conn.username) {
        return conn.username;
    }

    const connType = connTypeOrDefault(conn);
    switch (connType) {
        case "ssh":
            return "root";
        case "winrm":
            return "Administrator";
        default:
            throw new Error(`unrecognized connectiont ype ${connType}`);
    }
}

function connToSsh2(conn: pulumi.Unwrap<ConnectionArgs>): any {
    return {
        host: conn.host,
        port: conn.port,
        username: connUsernameOrDefault(conn),
        password: conn.password,
        privateKey: conn.privateKey,
        passphrase: conn.privateKeyPassphrase,
    };
}

function copyFile(conn: pulumi.Unwrap<ConnectionArgs>, src: string, dest: string): Promise<void> {
    const connType = connTypeOrDefault(conn);
    if (connType !== "ssh") {
        throw new Error("only SSH connection types currently supported");
    }

    const scp2 = require("scp2");
    let connectionFailCount = 0;
    return new Promise((resolve, reject) => {
        function scp() {
            scp2.scp(
                src,
                { path: dest, ...connToSsh2(conn) },
                (err) => {
                    if (err) {
                        connectionFailCount++;
                        if (connectionFailCount > 10) {
                            reject(err);
                        } else {
                            setTimeout(scp, connectionFailCount * 500);
                        }
                        return;
                    }
                    resolve();
                },
            );
        }
        scp();
    });
}

function runCommand(conn: pulumi.Unwrap<ConnectionArgs>, cmd: string): Promise<string> {
    const connType = connTypeOrDefault(conn);
    if (connType !== "ssh") {
        throw new Error("only SSH connection types currently supported");
    }

    const ssh2 = require("ssh2");
    const sshConn = connToSsh2(conn);
    let connectionFailCount = 0;
    return new Promise((resolve, reject) => {
        const conn = new ssh2.Client();
        function connect() {
            conn.on("ready", () => {
                conn.exec(cmd, (err, stream) => {
                    if (err) {
                        reject(err);
                        return;
                    }
                    stream.on("close", (code, signal) => {
                        conn.end();
                        if (code) {
                            reject(new Error("Command exited with " + code));
                        } else {
                            resolve();
                        }
                    }).on("data", (data) => {
                        console.log(data.toString("utf8"));
                    }).stderr.on("data", (data) => {
                        console.error(data.toString("utf8"));
                    });
                });
            }).on("error", (err) => {
                connectionFailCount++;
                if (connectionFailCount > 10) {
                    reject(err);
                } else {
                    setTimeout(connect, connectionFailCount * 500);
                }
            }).connect(sshConn);
        }
        connect();
    });
}


// CopyFile is a provisioner step that can copy a file from the machine running Pulumi to the newly created resource.
export class CopyFile extends pulumi.ComponentResource {
    private readonly provisioner: Provisioner<ConnectionArgs>;

    constructor(name: string, args: CopyFileArgs, opts?: pulumi.ComponentResourceOptions) {
        super("pulumi:provisioners:CopyFile", name, args, opts);

        this.provisioner = new Provisioner<ConnectionArgs>(
            `${name}-provisioner`,
            {
                dep: args.conn,
                onCreate: (conn) => copyFile(conn, args.src, args.dest),
            },
            { parent: this, ...opts || {} },
        );
    }
}

export interface CopyFileArgs {
    // conn contains information on how to connect to the destination, in addition to dependency information.
    conn: pulumi.Input<ConnectionArgs>;
    // src is the source of the file or directory to copy. It can be specified as relative to the current
    // working directory or as an absolute path. This cannot be specified if content is set.
    src: string;
    // // content is the in-memory content to copy to the destination. If the destination is a file, the content
    // // will be written on that file, whereas if it's a directory, a file named `pulumi-content` is created.
    // content?: pulumi.Input<string>;
    // dest is required and specifies the absolute path on the target where the file will be copied to.
    dest: string;
}

// RemoteExec runs remote commands and/or invokes scripts. If commands and scripts are specified, they are
// run in the following order: command, commands, script, and finally then scripts.
export class RemoteExec extends pulumi.ComponentResource {
    private readonly provisioner: Provisioner<ConnectionArgs>;

    constructor(name: string, args: RemoteExecArgs, opts?: pulumi.ComponentResourceOptions) {
        super("pulumi:provisioners:RemoteExec", name, args, opts);

        this.provisioner = new Provisioner<ConnectionArgs>(
            `${name}-provisioner`,
            {
                dep: args.conn,
                onCreate: async (conn) => {
                    if (args.command) {
                        await runCommand(conn, args.command);
                    }
                    if (args.commands) {
                        for (const cmd of args.commands) {
                            await runCommand(conn, cmd);
                        }
                    }
                },
            },
            { parent: this, ...opts || {} },
        );
    }
}

export interface RemoteExecArgs {
    conn: ConnectionArgs;
    command?: string;
    commands?: string[];
}

File: aws-py-ec2-provisioners/provisioners.py
# Copyright 2020, Pulumi Corporation.  All rights reserved.

import abc
import json
import io
import paramiko
import pulumi
from pulumi import dynamic
import sys
import time
from typing import Any, Optional
from typing_extensions import TypedDict
from uuid import uuid4

# ConnectionArgs tells a provisioner how to access a remote resource. It includes the hostname
# and optional port (default is 22), username, password, and private key information.
class ConnectionArgs(TypedDict):
    host: pulumi.Input[str]
    """The host to SSH into."""
    port: Optional[pulumi.Input[int]] = None
    """The port to SSH into (default 22)."""
    username: Optional[pulumi.Input[str]] = None
    """The username for the SSH login."""
    password: Optional[pulumi.Input[str]] = None
    """The optional password for the SSH login (private key is recommended instead)."""
    private_key: Optional[pulumi.Input[str]] = None
    """The private key, as an ASCII string, to use for the SSH connection."""
    private_key_passphrase: Optional[pulumi.Input[str]] = None
    """The private key passphrase, if any, to use for the SSH private key."""

def connect(conn: ConnectionArgs) -> paramiko.SSHClient:
    ssh = paramiko.SSHClient()
    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    skey = io.StringIO(conn['private_key'])
    pkey = paramiko.RSAKey.from_private_key(skey, password=conn.get('private_key_passphrase'))

    # Retry the connection until the endpoint is available (up to 2 minutes).
    retries = 0
    while True:
        try:
            ssh.connect(
                hostname=conn['host'],
                port=conn.get('port') or 22,
                username=conn.get('username'),
                password=conn.get('password'),
                pkey=pkey,
            )
            return ssh
        except paramiko.ssh_exception.NoValidConnectionsError:
            if retries == 24:
                raise
            time.sleep(5)
            retries = retries + 1
            pass

class ProvisionerProvider(dynamic.ResourceProvider):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def on_create(self, inputs: Any) -> Any:
        return

    def create(self, inputs):
        outputs = self.on_create(inputs)
        return dynamic.CreateResult(id_=uuid4().hex, outs=outputs)

    def diff(self, _id, olds, news):
        # If anything changed in the inputs, replace the resource.
        diffs = []
        for key in olds:
            if key not in news:
                diffs.append(key)
            else:
                olds_value = json.dumps(olds[key], sort_keys=True, indent=2)
                news_value = json.dumps(news[key], sort_keys=True, indent=2)
                if olds_value != news_value:
                    diffs.append(key)
        for key in news:
            if key not in olds:
                diffs.append(key)

        return dynamic.DiffResult(changes=len(diffs) > 0, replaces=diffs, delete_before_replace=True)

# CopyFileProvider implements the resource lifecycle for the CopyFile resource type below.
class CopyFileProvider(ProvisionerProvider):
    def on_create(self, inputs: Any) -> Any:
        ssh = connect(inputs['conn'])
        scp = ssh.open_sftp()
        try:
            scp.put(inputs['src'], inputs['dest'])
        finally:
            scp.close()
            ssh.close()
        return inputs

# CopyFile is a provisioner step that can copy a file over an SSH connection.
class CopyFile(dynamic.Resource):
    def __init__(self, name: str, conn: pulumi.Input[ConnectionArgs],
                 src: str, dest: str, opts: Optional[pulumi.ResourceOptions] = None):
        self.conn = conn
        """conn contains information on how to connect to the destination, in addition to dependency information."""
        self.src = src
        """
        src is the source of the file or directory to copy. It can be specified as relative to the current
        working directory or as an absolute path. This cannot be specified if content is set.
        """
        self.dest = dest
        """dest is required and specifies the absolute path on the target where the file will be copied to."""

        super().__init__(
            CopyFileProvider(),
            name,
            {
                'dep': conn,
                'conn': conn,
                'src': src,
                'dest': dest,
            },
            opts,
        )

# RunCommandResult is the result of running a command.
class RunCommandResult(TypedDict):
    stdout: str
    """The stdout of the command that was executed."""
    stderr: str
    """The stderr of the command that was executed."""

# RemoteExecProvider implements the resource lifecycle for the RemoteExec resource type below.
class RemoteExecProvider(ProvisionerProvider):
    def on_create(self, inputs: Any) -> Any:
        ssh = connect(inputs['conn'])
        try:
            results = []
            for command in inputs['commands']:
                stdin, stdout, stderr = ssh.exec_command(command)
                results.append({
                    'stdout': ''.join(stdout.readlines()),
                    'stderr': ''.join(stderr.readlines()),
                })
            inputs['results'] = results
            print(f'results: {results}')
        finally:
            ssh.close()
        return inputs

# RemoteExec runs remote one or more commands over an SSH connection. It returns the resulting
# stdout and stderr from the commands in the results property.
class RemoteExec(dynamic.Resource):
    results: pulumi.Output[list]

    def __init__(self, name: str, conn: ConnectionArgs, commands: list, opts: Optional[pulumi.ResourceOptions] = None):
        self.conn = conn
        """conn contains information on how to connect to the destination, in addition to dependency information."""
        self.commands = commands
        """The commands to execute. Exactly one of 'command' and 'commands' is required."""
        self.results = []
        """The resulting command outputs."""

        super().__init__(
            RemoteExecProvider(),
            name,
            {
                'conn': conn,
                'commands': commands,
                'results': None,
            },
            opts,
        )
